src/hotspot/share/opto/addnode.cpp

```diff
@@ -769,6 +771,553 @@ uint AddPNode::match_edge(uint idx) const {
+  MergeLoadInfo* merge_load_info(LoadNode* load) const;
+  // Make the merged load and optional ReverseByte for replace
+  Node* make_merged_load(const MergeLoadInfoList& merge_list);
+
+  // Helper methods for merge loads optimization
+  static bool is_supported_load_opcode(int opcode);
+  static bool is_supported_combine_opcode(int opcode);
+  // Check the two combine operator Nodes can be reachable from one to another, and intermediate nodes has the same type
+  static bool is_reachable_combine_nodes(const Node* from, const Node* to);
+  MemoryAdjacentStatus get_adjacent_load_status(const LoadNode* first, const LoadNode* second) const;
+  // Go through ConvI2L which is unique output of input node
+  static const Node* by_pass_i2l(const Node* n);
+
+#ifndef PRODUCT
+  // Access to TraceMergeLoads tags
+  bool is_trace(TraceMergeLoads::Tag tag) const {
+    return _trace_tags.at(tag);
+  }
+
+  bool is_trace_basic() const {
+    return is_trace(TraceMergeLoads::Tag::BASIC);
+  }
+
+  bool is_trace_pointer_parsing() const {
+    return is_trace(TraceMergeLoads::Tag::POINTER_PARSING);
+  }
+
+  bool is_trace_pointer_aliasing() const {
+    return is_trace(TraceMergeLoads::Tag::POINTER_ALIASING);
+  }
+
+  bool is_trace_pointer_adjacency() const {
+    return is_trace(TraceMergeLoads::Tag::POINTER_ADJACENCY);
+  }
+
+  bool is_trace_success() const {
+    return is_trace(TraceMergeLoads::Tag::SUCCESS);
+  }
+#endif
+};
+
+bool MergePrimitiveLoads::is_supported_load_opcode(int opc) {
+  // Check for B/S/C/I
+  if (opc != Op_LoadB && opc != Op_LoadUB &&
+      opc != Op_LoadS && opc != Op_LoadUS &&
+      opc != Op_LoadI) {
+    return false;
+  }
+  return true;
+}
+
+bool MergePrimitiveLoads::is_supported_combine_opcode(int opc) {
+  return opc == Op_OrI || opc == Op_OrL;
+}
+
+// Go through ConvI2L which is unique output of input node
+const Node* MergePrimitiveLoads::by_pass_i2l(const Node* n) {
+  if (n != nullptr && n->outcnt() == 1 && n->unique_out()->Opcode() == Op_ConvI2L) {
+    return n->unique_out();
+  } else {
+    return n;
+  }
+};
+
+/*
+ * Check the _combine operator can be a candidate to trigger merge loads.
+ * The candidate load has the pattern:
+ *
+ * Load -> OrI/OrL
+ *  It has no LShift usage and the And node is optimized out in previous optimization
+ *
+ *  The reference java code snippets are:
+ *              ...
+ *            | (((long) array[offset + 5] & 0xff) << 16)
+ *            | (((long) array[offset + 6] & 0xff) << 8 )
+ *            | (((long) array[offset + 7] & 0xff)      );                     <----- candidate
+ *
+ *              ((UNSAFE.getByte(array, address    ) & 0xff)      )            <----- candidate
+ *            | ((UNSAFE.getByte(array, address + 1) & 0xff) <<  8)
+ *            | ((UNSAFE.getByte(array, address + 2) & 0xff) << 16)
+ *            | ((UNSAFE.getByte(array, address + 3) & 0xff) << 24);
+ */
+bool MergePrimitiveLoads::is_merged_load_candidate() const {
```

Comment:
What is your definition of candidate here? It seems to have something to do about not having a `LShift`, why? Maybe you can give a good definition here or somewhere else?

Reply:
Comments is updated. Only the last one for combine operator chain is the candadte.